### PR TITLE
tests: use testing package to run tests

### PR DESCRIPTION
**Pull request by shantuo**
* extract commmon test functions to internal/testutil package

### CODE REVIEW

**Comment by zombiezen** [comment]
One larger observation (@ijt may also want to weigh in):

These tests are not hermetic, in that they must be set up via the shell script runner. It seems like this would complicate the `go test ./...` story. Perhaps what we should do is if the tests don't have the flags set that they are skipped?

**Comment by ijt** [comment]
@zombiezen, not sure I follow. I thought we were going to rely on `-record` for that. Shall we chat in person tomorrow?

**Comment by zombiezen** [comment]
This test is a slightly different category in that record/replay doesn't make much sense for this one.

File: tests/internal/testutil/testutil.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	maxBackoff = time.Minute
+)
+
+// Retry keeps retrying hitting url using f with exponentially increased delay
+// until the delay passes maxBackoff.
+func Retry(t *testing.T, s string, f func(string) error) {
+	var err error
+	for d := startDelay; d <= maxBackoff; d *= 2 {
+		time.Sleep(d)
+		if err = f(s); err == nil {
```
[the remark this answers is not in the document]
Comment (zombiezen) [reply tests/internal/testutil/testutil.go]: There are some errors (bad configuration etc.) where retrying will not help. However, it mostly was around creating the client, so now that you've moved that out of the retry loop, I'm not worried about this anymore.

File: tests/internal/testutil/testutil.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+)
+
+const (
+	startDelay = 5 * time.Second
+	maxBackoff = time.Minute
+)
+
+// Retry keeps retrying hitting url using f with exponentially increased delay
+// until the delay passes maxBackoff.
+func Retry(t *testing.T, s string, f func(string) error) {
```
[the remark this answers is not in the document]
Comment (zombiezen) [reply tests/internal/testutil/testutil.go]: You could instead rely on the functions closing on the variables they need.

File: tests/gcp/app/app_test.go [hunk 3]
```diff
@@ -33,62 +33,87 @@ const (
 )
 
 var (
-	address   string
-	projectID string
+	address        string
+	projectID      string
+	logadminClient *logadmin.Client
```
Comment (zombiezen) [inline tests/gcp/app/app_test.go]: Don't pass these as global variables. See my remarks about the `Retry` function.

File: tests/internal/testutil/testutil.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	startDelay = 5 * time.Second
+	maxBackoff = time.Minute
+)
+
+// Logger writes logs (normally error messages) during retries.
+type Logger interface {
+	Logf(format string, args ...interface{})
+}
+
+// Retry keeps retrying hitting url using f with exponentially increased delay
```
Comment (zombiezen) [inline tests/internal/testutil/testutil.go]: This comment needs updating:

Retry calls f until it no longer returns an error, waiting between tries.

File: tests/internal/testutil/testutil.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		return err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("error response got: %s", resp.Status)
+	}
+	return err
+}
+
+// URLSuffix append a human-readible suffix to the test URL.
+func URLSuffix(addr string) string {
```
[the remark this answers is not in the document]
Comment (zombiezen) [reply tests/internal/testutil/testutil.go]: This function is more composable by removing the append behavior. Just let this be an opaque value returned.

**Comment by zombiezen** [comment]
Heads-up: for better or for worse, I replied to old comment threads, so you may have to wade through the clunky pull request UI.

File: tests/internal/testutil/testutil.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (shantuo) [reply tests/internal/testutil/testutil.go]: Done.

File: tests/internal/testutil/testutil.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (shantuo) [reply tests/internal/testutil/testutil.go]: Ack.

File: tests/internal/testutil/testutil.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (shantuo) [reply tests/internal/testutil/testutil.go]: Done.

**Pull request merged**